src/rust/Dockerfile.Makefile

```diff
@@ -0,0 +1,214 @@
+        musl-dev \
+        musl-tools \
+        wait-for-it \
+        wget \
+        zip \
+    && rm -rf /var/lib/apt/lists/*
+
+RUN adduser \
+        --disabled-password \
+        --gecos '' \
+        --uid 19999 \
+        --home /home/grapl \
+        --shell /bin/bash \
+        grapl
+
+USER grapl
+ENV CARGO_HOME=/home/grapl/.cargo
+WORKDIR /home/grapl
+
+RUN mkdir /home/grapl/.cargo; \
+    if test "${TARGET}" = "x86_64-unknown-linux-musl"; then \
+      rustup target add x86_64-unknown-linux-musl; \
+    fi
+
+# copy sources
+COPY --chown=grapl . .
+
+#
+# build using docker volume for cache
+#
+FROM base AS build-sccache-none
+
+RUN if test "${PROFILE}" = "release"; then \
+      cargo build --target=${TARGET} --release; \
+    elif test "${PROFILE}" = "debug"; then \
+      cargo build --target=${TARGET}; \
+    else \
+      echo "ERROR: Unknown build profile: ${PROFILE}"; \
+      exit 1; \
+    fi
+
+#
+# sccache base
+#
+FROM base AS sccache
+
+ENV SCCACHE_DIR=/home/grapl/sccache
+ENV RUSTC_WRAPPER=/home/grapl/bin/sccache
+
+# Waiting on the following sccache PR to land to better support S3:
+# https://github.com/mozilla/sccache/pull/869
+RUN if test "${SCCACHE_LOCATION}" != "s3" ; then \
+      cd /tmp && \
+      wget -q https://github.com/mozilla/sccache/releases/download/0.2.14/sccache-0.2.14-x86_64-unknown-linux-musl.tar.gz && \
+      tar xvzf sccache-0.2.14-x86_64-unknown-linux-musl.tar.gz && \
+      mkdir -p /home/grapl/bin && \
+      cp sccache-0.2.14-x86_64-unknown-linux-musl/sccache /home/grapl/bin/sccache; \
+    fi
+
+# RUN cd /tmp && \
+#     wget -q https://github.com/mozilla/sccache/releases/download/0.2.14/sccache-0.2.14-x86_64-unknown-linux-musl.tar.gz && \
+#     tar xvzf sccache-0.2.14-x86_64-unknown-linux-musl.tar.gz && \
+#     mkdir /home/grapl/bin && \
+#     cp sccache-0.2.14-x86_64-unknown-linux-musl/sccache /home/grapl/bin/sccache
+
+
+# use docker volume cache for sccache
+FROM sccache AS build-sccache-container
+
+RUN --mount=type=cache,uid=19999,gid=19999,target=/home/grapl/.cargo/registry \
+    --mount=type=cache,uid=19999,gid=19999,target=/home/grapl/sccache \
+    if test "${PROFILE}" = "release"; then \
+      cargo build --target=${TARGET} --release; \
+    elif test "${PROFILE}" = "debug"; then \
+      cargo build --target=${TARGET}; \
+    else \
+      echo "ERROR: Unknown build profile: ${PROFILE}"; \
+      exit 1; \
+    fi
+
+# build-sccache-host is not currently supported
+# FROM sccache AS build-sccache-host
+
+# build using sccache from S3
+FROM sccache AS build-sccache-s3
+
+ARG SCCACHE_BUCKET
+
+ENV SCCACHE_BUCKET=${SCCACHE_BUCKET}
+ENV SCCACHE_S3_KEY_PREFIX=sccache
+
+SHELL ["/bin/bash", "-c"]
+
+RUN --mount=type=cache,uid=19999,gid=19999,target=/home/grapl/.cargo/registry \
+    --mount=type=cache,uid=19999,gid=19999,target=/home/grapl/sccache \
+    --mount=type=secret,uid=19999,gid=19999,id=aws,dst=/home/grapl/awscreds \
```

Comment:
(what's up with these hardcoded ids?)

Reply:
> In all honesty I don't have the full background to give this an approve or not, but it seems reasonably thought out.
> 
> Notably, I don't see any e2e tests #justsayin
> 
> if you're like "this PR is big enough already" then i'm down with that, strip out dobi etc in a followup is ok

Originally I was just thinking this would sit side-by-side to dobi while we test it out and get a feel for it. I didn't want to port everything  over yet in case we had other or new ideas for direction or something, but I think it looks promising so I'm open to doing it in just one PR.

Reply:
> (what's up with these hardcoded ids?)

Looked like I couldn't reference a user by name here*. I didn't want to run commands as root and thought running as a UID that probably doesn't map to host would be a good idea, so .. that's what I came up with.

* https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/syntax.md#run---mounttypecache